### PR TITLE
feat: codegen for Value Entity Handler

### PR DESCRIPTION
**Pull request by octonato**

### CODE REVIEW

File: codegen/java-gen/src/main/scala/com/lightbend/akkasls/codegen/java/EntityServiceSourceGenerator.scala [hunk 1]
```diff
@@ -148,7 +148,7 @@ object EntityServiceSourceGenerator {
       case eventSourcedEntity: EventSourcedEntity =>
         eventSourcedEntitySource(service, eventSourcedEntity, packageName, className, interfaceClassName, entityType)
       case valueEntity: ValueEntity =>
-        valueEntitySource(service, valueEntity, packageName, className, interfaceClassName, entityType)
+        ValueEntitySourceGenerator.valueEntitySource(service, valueEntity, packageName, className, interfaceClassName)
```
Comment (octonato) [inline codegen/java-gen/src/main/scala/com/lightbend/akkasls/codegen/java/EntityServiceSourceGenerator.scala:151]: Starting to move thing out to its own generator. For the moment just calling the method in new `ValueEntitySourceGenerator`

File: sdk/src/main/java/com/akkaserverless/javasdk/lowlevel/ValueEntityHandler.java [hunk 2]
```diff
@@ -37,8 +37,8 @@
    * @param context The command context.
    * @return The reply to the command, if the command isn't being forwarded elsewhere.
    */
-  ValueEntityBase.Effect<Any> handleCommand(Any command, Any state, CommandContext<Any> context)
-      throws Throwable;
+  ValueEntityBase.Effect<? extends GeneratedMessageV3> handleCommand(
+      Any command, Any state, CommandContext<Any> context) throws Throwable;
```
Comment (octonato) [inline sdk/src/main/java/com/akkaserverless/javasdk/lowlevel/ValueEntityHandler.java:41]: this is a leftover from previous PR #73 

I will move this to separate PR later, should not be here.

File: sdk/src/test/scala/com/akkaserverless/javasdk/impl/valueentity/AnnotationBasedValueEntitySupportSpec.scala [hunk 3]
```diff
@@ -102,6 +104,11 @@ class AnnotationBasedValueEntitySupportSpec extends AnyWordSpec with Matchers {
       case UpdateState(any) => any.asInstanceOf[S]
     }
 
+  def decodeWrapped(anyMessage: GeneratedMessageV3): Wrapped = {
+    // TODO: which extra check could we do here?
+    WrappedResolvedType.parseFrom(anyMessage.toByteString)
```
Comment (octonato) [inline sdk/src/test/scala/com/akkaserverless/javasdk/impl/valueentity/AnnotationBasedValueEntitySupportSpec.scala:109]: will move to separate PR

File: samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java [hunk 4]
```diff
@@ -16,118 +16,81 @@
...
-      } else {
-        unwrapped = e;
-      }
-      if (unwrapped.getClass().isAssignableFrom(FailInvoked$.class)) {
-        throw unwrapped;
+      if (e.getClass().isAssignableFrom(FailInvoked$.class)) {
+        throw e;
       } else {
-        throw unwrapped.getCause();
+        throw e.getCause();
```
Comment (octonato) [inline samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java:57]: @raboof, I'm trying to reduce the size of this code. 

My understanding is that InvocationTargetException is a leftover from a reflection based API. We don't need it here anymore. 

So, at the end, either we have a `FailInvoked` that we should throw as is or we have something else in which case we should throw the cause of it. 

Or maybe we should not have any try/catch here and just let whatever exception we get propagate further? WDYT?

**Review by octonato — COMMENTED** [review]
Still incomplete and very draft. 

Putting it out so I can switch to PR review for now and eventually get some early feedback.

File: samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java (hunk 4 above)
Comment (raboof) [reply samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java:57]: > My understanding is that InvocationTargetException is a leftover from a reflection based API. We don't need it here anymore

Sounds plausible to me.

> maybe we should not have any try/catch here and just let whatever exception we get propagate further? WDYT?

So you want to move the handling/unwrapping of `FailInvoked` to somewhere else? I have no objection ;)

Comment (octonato) [reply samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java:57]: no, my idea is to simple not have the try/catch. We don't unwrap `FailInvoked`, we just rethrow it as is. And for other exceptions we throw the cause, but I wonder if it makes sense. We may just let it bubble up for whatever we get.

File: codegen/java-gen/src/main/scala/com/lightbend/akkasls/codegen/java/ValueEntitySourceGenerator.scala [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+          |      return invoke(command, state, context);
+          |    } catch (Exception e) {
+          |      if (e.getClass().isAssignableFrom(FailInvoked$$.class)) {
+          |        throw e;
+          |      } else {
+          |        throw e.getCause();
+          |      }
+          |    }
+          |  }
+          |}""".stripMargin
```
Comment (octonato) [inline codegen/java-gen/src/main/scala/com/lightbend/akkasls/codegen/java/ValueEntitySourceGenerator.scala:152]: BTW, this is what a mean by writing a codegen where we can recognize the generated code directly instead of continually have to parse `<>`, `<+>`, `line` and all that syntax.

File: samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java (hunk 4 above)
Comment (johanandren) [reply samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java:57]: Aren't we getting rid of FailInvoked from the context.fail anyway, and failing is rather done with `effects().error()`?

Comment (raboof) [reply samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java:57]: Of course! :+1:

Comment (octonato) [reply samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java:57]: but a user can always throw an exception, no? 

In any case, as suggested in #151, I think we should not have any error handling at all in that part of the code.

**Comment by octonato** [comment]
value entity shopping cart is now using the new generation, but we get a runtime error. Will fix tomorrow...

```scala
java.lang.ClassCastException: class com.google.protobuf.Empty cannot be cast to class com.google.protobuf.Any (com.google.protobuf.Empty and com.google.protobuf.Any are in unnamed module of loader 'app')
        at com.akkaserverless.javasdk.impl.effect.EffectSupport$.asProtocol(EffectSupport.scala:37)
        at com.akkaserverless.javasdk.impl.AbstractClientActionContext.replyToClientAction(Contexts.scala:133)
        at com.akkaserverless.javasdk.impl.AbstractClientActionContext.replyToClientAction$(Contexts.scala:120)
        at com.akkaserverless.javasdk.impl.valueentity.ValueEntitiesImpl$CommandContextImpl.replyToClientAction(ValueEntitiesImpl.scala:216)
        at com.akkaserverless.javasdk.impl.valueentity.ValueEntitiesImpl.$anonfun$runEntity$3(ValueEntitiesImpl.scala:165)
        at akka.stream.impl.fusing.Scan$$anon$11.onPush(Ops.scala:442)
        at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:542)
        at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:423)
        at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:650)
        at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:521)
        at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:625)
        at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:800)
        at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$shortCircuitBatch(ActorGraphInterpreter.scala:787)
        at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:819)
        at akka.actor.Actor.aroundReceive(Actor.scala:537)
        at akka.actor.Actor.aroundReceive$(Actor.scala:535)
        at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:716)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:580)
        at akka.actor.ActorCell.invoke(ActorCell.scala:548)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
        at akka.dispatch.Mailbox.run(Mailbox.scala:231)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
```

File: codegen/java-gen/src/main/scala/com/lightbend/akkasls/codegen/java/ValueEntitySourceGenerator.scala [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+      interfaceClassName: String
+  ): Document = {
+    val messageTypes = service.commands.toSeq
+        .flatMap(command => Seq(command.inputType, command.outputType)) ++ Seq(entity.state.fqn)
+
+    val imports = (messageTypes
+      .filterNot(_.parent.javaPackage == packageName)
+      .map(typeImport) ++
+    Seq(
+      "com.akkaserverless.javasdk.valueentity.ValueEntityContext"
```
Comment (raboof) [inline codegen/java-gen/src/main/scala/com/lightbend/akkasls/codegen/java/ValueEntitySourceGenerator.scala:41]: This one doesn't seem needed anymore, either

File: codegen/java-gen/src/main/scala/com/lightbend/akkasls/codegen/java/ValueEntitySourceGenerator.scala [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+      Some(entity.state),
+      packageName,
+      otherImports = Seq(
+        "com.akkaserverless.javasdk.impl.AnySupport",
+        "com.akkaserverless.javasdk.impl.EntityExceptions",
+        "com.akkaserverless.javasdk.impl.valueentity.AdaptedCommandContextWithState",
+        "com.akkaserverless.javasdk.lowlevel.ValueEntityHandler",
+        "com.akkaserverless.javasdk.valueentity.CommandContext",
+        "com.akkaserverless.javasdk.valueentity.ValueEntityBase",
+        "com.akkaserverless.javasdk.valueentity.ValueEntityContext",
```
Comment (raboof) [inline codegen/java-gen/src/main/scala/com/lightbend/akkasls/codegen/java/ValueEntitySourceGenerator.scala:121]: here, neither

**Review by raboof — APPROVED** [review]

File: sdk/src/main/scala/com/akkaserverless/javasdk/impl/valueentity/ValueEntitiesImpl.scala [hunk 8]
```diff
@@ -164,11 +165,23 @@ final class ValueEntitiesImpl(_system: ActorSystem,
...
+            case UpdateState(state) => UpdateState(service.anySupport.encodeJava(state))
+            case DeleteState => DeleteState
+            case NoPrimaryEffect => NoPrimaryEffect
+          }
+          // FIXME something more that needs to be serialized?
+          val serializedSecondaryEffect = effect.secondaryEffect match {
+            case MessageReplyImpl(message, metadata, sideEffects) =>
+              MessageReplyImpl(service.anySupport.encodeJava(message), metadata, sideEffects)
+            case other => other
+          }
```
Comment (octonato) [inline sdk/src/main/scala/com/akkaserverless/javasdk/impl/valueentity/ValueEntitiesImpl.scala:178]: This is something that was causing the integration tests to fail. 

We must convert whatever type we get to PbAny and that was not always the case.

**Review by octonato — COMMENTED** [review]
@raboof, I removed the unused imports and got the integration tests for the counter value entity fixed. 

I also squashed it, which I realise now that was a bad idea as it make harder for you to see my latest changes. Sorry. :-)

File: .circleci/config.yml [hunk 9]
```diff
@@ -209,15 +209,14 @@ jobs:
...
             # mvn -Dakkaserverless-sdk.version=$SDK_VERSION verify -Pit
             mvn -Dakkaserverless-sdk.version=$SDK_VERSION verify
       - run: 
           name: Run integration tests for Counter Value Entity sample
           command: |
             cd samples/valueentity-counter
-            echo "Running mvn with SDK version: '$SDK_VERSION' - FIXME disabled for codegen/API work"
-            # mvn -Dakkaserverless-sdk.version=$SDK_VERSION verify -Pit
+            echo "Running mvn with SDK version: '$SDK_VERSION'"
+            mvn -Dakkaserverless-sdk.version=$SDK_VERSION verify -Pit
```
Comment (raboof) [inline .circleci/config.yml:219]: \o/ \o/

File: codegen/java-gen/src/test/scala/com/lightbend/akkasls/codegen/java/EntityServiceSourceGeneratorSuite.scala [hunk 10]
```diff
@@ -19,6 +19,7 @@ package java
 
 class EntityServiceSourceGeneratorSuite extends munit.FunSuite {
 
+  // -- EventSourcedEntity
```
Comment (raboof) [inline codegen/java-gen/src/test/scala/com/lightbend/akkasls/codegen/java/EntityServiceSourceGeneratorSuite.scala:22]:
```suggestion
```

File: samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java [hunk 11]
```diff
@@ -28,70 +28,59 @@
...
-  CustomerValueEntityHandler(CustomerValueEntity entity) {
-    this.entity = entity;
-    this.anySupport =
-        new AnySupport(
-            new Descriptors.FileDescriptor[] {CustomerDomain.getDescriptor()},
-            CustomerValueEntityHandler.class.getClassLoader(),
-            AnySupport.DefaultTypeUrlPrefix(),
-            AnySupport.PREFER_JAVA());
+  CustomerValueEntityHandler() {
+    this.entity = new CustomerValueEntity();
```
Comment (raboof) [inline samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java:47]: I'm not sure we should construct the user implementation from the generated code: that would mean the user no longer has the opportunity to pass in additional utilities that might be needed in their implementation methods

File: samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java [hunk 12]
```diff
@@ -28,70 +28,59 @@
...
   final CustomerValueEntity entity;
-  final AnySupport anySupport;
 
   public static final Descriptors.ServiceDescriptor serviceDescriptor =
       CustomerApi.getDescriptor().findServiceByName("CustomerService");
   public static final String entityType = "customers";
 
-  CustomerValueEntityHandler(CustomerValueEntity entity) {
-    this.entity = entity;
-    this.anySupport =
```
Comment (raboof) [inline samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java:49]: very nice, indeed no longer needed because we no longer allow updating the state via the context :+1:

File: sdk/src/main/scala/com/akkaserverless/javasdk/impl/valueentity/ValueEntitiesImpl.scala (hunk 8 above)
Comment (raboof) [reply sdk/src/main/scala/com/akkaserverless/javasdk/impl/valueentity/ValueEntitiesImpl.scala:178]: Right, so here we choose to do the serialization in the `Impl` rather than in the `Handler`. Let's see how long we can get away with that :)

File: sdk/src/main/scala/com/akkaserverless/javasdk/impl/valueentity/ValueEntitiesImpl.scala [hunk 13]
```diff
@@ -164,11 +165,23 @@ final class ValueEntitiesImpl(_system: ActorSystem,
...
+
           val clientAction =
-            context.replyToClientAction(effect.secondaryEffect, allowNoReply = false, restartOnFailure = false)
+            context.replyToClientAction(serializedSecondaryEffect, allowNoReply = false, restartOnFailure = false)
 
-          if (!context.hasError && !effect.secondaryEffect.isInstanceOf[ErrorReplyImpl[_]]) {
-            val (nextState: Option[ScalaPbAny], action: Option[ValueEntityAction]) = effect.primaryEffect match {
+          if (!context.hasError && !serializedSecondaryEffect.isInstanceOf[ErrorReplyImpl[_]]) {
+            val (nextState, action) = serializedPrimaryEffect match {
```
Comment (raboof) [inline sdk/src/main/scala/com/akkaserverless/javasdk/impl/valueentity/ValueEntitiesImpl.scala:184]: aren't those type annotations nice to have? I remember I was having trouble keeping track of where things were serialized and where they were not ;)

File: samples/java-customer-registry/src/main/java/customer/Main.java [hunk 14]
```diff
@@ -52,7 +52,7 @@ public static void main(String[] args) throws Exception {
           // end::register[]
           .lowLevel()
           .registerValueEntity(
-              context -> new CustomerValueEntityHandler(new CustomerValueEntity()),
+              context -> new CustomerValueEntityHandler(),
```
Comment (raboof) [inline samples/java-customer-registry/src/main/java/customer/Main.java:55]: Wouldn't it be convenient to give the user an opportunity to pass additional things to their `CustomerValueEntity` here?

**Review by raboof — COMMENTED** [review]
Not sure about that handler constructor, otherwise looking good

File: sdk/src/main/scala/com/akkaserverless/javasdk/impl/valueentity/ValueEntitiesImpl.scala (hunk 13 above)
Comment (octonato) [reply sdk/src/main/scala/com/akkaserverless/javasdk/impl/valueentity/ValueEntitiesImpl.scala:184]: I think I removed it because of some warnings. I will check and re-introduce if that's not the case.

File: samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java (hunk 11 above)
Comment (raboof) [reply samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java:47]: Perhaps we can provide some more utilities there, like a registration component that accepts a number of factories/lambdas to create the entities - but let's take one thing at a time ;)

Comment (octonato) [reply samples/java-customer-registry/src/main/java/customer/CustomerValueEntityHandler.java:47]: @franciscolopezsancho, this is something that affects the testkit. In #149 you asked if the testkit should init the entity by itself or not. My argument was that it was convenient to do so, but Arnout has an import point here. Users may want to define their constructor the way they want and pass things to it. 

So, both handler and testkit should receive an entity as parameter.

File: sdk/src/main/scala/com/akkaserverless/javasdk/impl/valueentity/ValueEntitiesImpl.scala (hunk 13 above)
Comment (octonato) [reply sdk/src/main/scala/com/akkaserverless/javasdk/impl/valueentity/ValueEntitiesImpl.scala]: added back

File: samples/java-customer-registry/src/main/java/customer/Main.java (hunk 14 above)
Comment (octonato) [reply samples/java-customer-registry/src/main/java/customer/Main.java]: constructor with entity is back

**Review by raboof — APPROVED** [review]
